Add GitHub dispatch provider (#369)
fluxcd/notification-controller: docs/spec/v1beta1/provider.md

```

Assuming that we deploy all Flux kustomization resources in the same namespace, it will be useful to have a unique kustomization resource name for each application in each environment. This will allow you to use only `event_type` to trigger tests for a exact application in the desired environment.

Let's say we have following folder structure for applications kustomization manifests:

```bash
app1/
    overlays/
        staging/
        production/
app2/
    overlays/
        staging/
        production/
```

You can then create a flux kustomization resource for each overlay to have unique `event_type` looks like this, for example, for app1 and the staging overlay: `Kustomization/app1-staging.default`.

The kustomization manifest for app1/staging:

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: app1-staging
  namespace: default
  spec:
    path: "./app1/staging"
```

You can the trigger tests for app1 in a staging environment when the app1 resources defined in `./app1/staging/` are configured by Flux:

```yaml
name: test-github-dispatch-provider
on:
  repository_dispatch:
    types:
      - "*app1-staging*"
```

> If the information from `event_type` is not enough to get the necessary information about the configured resources, you can get more information from `client_payload`. For example, you can get the message from the payload in the workfow job step:

Reviewer: Can you please change this to show how `summary` can be used.

Author: I have modified the examples in relation to our discussion in slack. The `summary` field from the webhook payload can be used to define the cluster on which the application is deployed.

--from-literal=address="Endpoint=sb://fluxv2.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=yoursaskeygeneatedbyazure"
```

### GitHub repository dispatch

> The `githubdispatch` webhook triggers a GitHub webhook event called [`repository_dispatch`](https://docs.github.com/en/rest/reference/repos#create-a-repository-dispatch-event) for the selected repository. The `repository_dispatch` event can be used to trigger a GitHub Actions workflow with tests for the deployed service.

Reviewer: ```suggestion
The `githubdispatch` provider generates GitHub  events of type [`repository_dispatch`](https://docs.github.com/en/rest/reference/repos#create-a-repository-dispatch-event) for the selected repository. The `repository_dispatch` events can be used to trigger GitHub Actions workflow.
```

The request includes the `event_type` and `client_payload` fields:

> * The `event_type` is generated by GitHub Dispatch provider by combining the Kind, Name and Namespace of the involved object in the format `{Kind}/{Name}.{Namespace}`. For example, the `event_type` for Flux configuration named `podinfo` in the `flux-system` namespace looks like this: `Kustomization/podinfo.flux-system`

Reviewer: ```suggestion
* The `event_type` is generated by GitHub Dispatch provider by combining the Kind, Name and Namespace of the involved object in the format `{Kind}/{Name}.{Namespace}`. For example, the `event_type` for a Flux Kustomization named `podinfo` in the `flux-system` namespace looks like this: `Kustomization/podinfo.flux-system`.
```

> * The `client_payload` contains the event data from the notification controller and looks like this:

Reviewer: ```suggestion
* The `client_payload` contains the Kubernetes event issued by Flux, e.g.:
```

```yaml
{
  involvedObject: {
    apiVersion: kustomize.toolkit.fluxcd.io/v1beta2,
    kind: Kustomization,
    name: podinfo,
    namespace: flux-system,
    resourceVersion: 426573,
    uid: b9b8554d-be26-4a3d-a97f-65f3276a097a
  },
  message: Deployment/podinfo/podinfo configured,
  metadata: {
    revision: main/96139968ca46b53462d1bf94de410a811d2026a1,
    summary: "staging (us-west-2)"
  },
  reason: Progressing,
  reportingController: kustomize-controller,
  reportingInstance: kustomize-controller-79464d8dc5-nb9c4,
  severity: info,
  timestamp: 2022-04-20T12:20:28Z
}
```

#### Setting up a GitHub workflow

> You need to add `repository_dispatch` trigger to trigger your GitHub workflow using the Flux notification with the relevant event type:

Reviewer: To trigger a GitHub Actions workflow when a Flux Kustomization finishes reconciling, you need to set the event type on the `repository_dispatch` to match the Flux object ID:

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: app1
  namespace: flux-system
  spec:
    path: "./app1/staging"
```

You would also like to know from the notification which cluster is being used for deployment. You can add the `spec.summary` field to the Flux alert configuration to mention the relevant cluster:

```yaml
apiVersion: notification.toolkit.fluxcd.io/v1beta1
kind: Alert
metadata:
  name: github-dispatch
  namespace: flux-system
spec:
  summary: "staging (us-west-2)"
  providerRef:
    name: github-dispatch
  eventSeverity: info
  eventSources:
    - kind: Kustomization
      name: 'podinfo'
```

Now you can the trigger tests in the GitHub workflow for app1 in a staging cluster when the app1 resources defined in `./app1/staging/` are configured by Flux:

```yaml
name: test-github-dispatch-provider
on:
  repository_dispatch:
    types: [Kustomization/podinfo.flux-system]
jobs:
  run-tests-staging:
    if: github.event.client_payload.metadata.summary == 'staging (us-west-2)'
    runs-on: ubuntu-18.04
    steps:
    - name: Run tests
      run: echo "running tests.."
```

> ### Setting up the GitHub dispatch provider

Reviewer: Let's move this before "Setting up a GitHub workflow" please.

Author: Fixed
